framework/src/modules/bft/bft_votes.ts

```diff
@@ -0,0 +1,173 @@
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { BlockHeader } from '../../node/state_machine';
+import { BFTParametersCache } from './bft_params';
+import { BFTVotes } from './schemas';
+import { getBlockBFTProperties } from './utils';
+
+export const insertBlockBFTInfo = (
+	bftVotes: BFTVotes,
+	header: BlockHeader,
+	maxLength: number,
+): void => {
+	// eslint-disable-next-line no-param-reassign
+	bftVotes.blockBFTInfos = [getBlockBFTProperties(header), ...bftVotes.blockBFTInfos].slice(
+		0,
+		maxLength,
+	);
+};
+
+export const getHeightNotPrevoted = (bftVotes: BFTVotes): number => {
+	const [newBlockBFTInfo] = bftVotes.blockBFTInfos;
+	const { height: currentHeight } = newBlockBFTInfo;
+	let heightPreviousBlock = newBlockBFTInfo.maxHeightGenerated;
+
+	// iterate over blockBFTInfo objects in decreasing order by height
+	while (currentHeight - heightPreviousBlock < bftVotes.blockBFTInfos.length) {
+		const blockBFTInfo = bftVotes.blockBFTInfos[currentHeight - heightPreviousBlock];
+		if (
+			!blockBFTInfo.generatorAddress.equals(newBlockBFTInfo.generatorAddress) ||
+			blockBFTInfo.maxHeightGenerated >= heightPreviousBlock
+		) {
+			return heightPreviousBlock;
+		}
+		heightPreviousBlock = blockBFTInfo.maxHeightGenerated;
+	}
+	// return one less than the height of the blockBFTInfo with smallest height
+	return bftVotes.blockBFTInfos[bftVotes.blockBFTInfos.length - 1].height - 1;
+};
+
+export const updatePrevotesPrecommits = async (
+	bftVotes: BFTVotes,
+	paramsCache: BFTParametersCache,
+): Promise<void> => {
+	// when processing the genesis block the array is empty and no prevotes or precommits need to be updated
+	if (bftVotes.blockBFTInfos.length === 0) {
+		return;
+	}
+	const [newBlockBFTInfo] = bftVotes.blockBFTInfos;
+	// a block header only implies votes if maxHeightGenerated < height
+	if (newBlockBFTInfo.maxHeightGenerated >= newBlockBFTInfo.height) {
+		return;
+	}
+	const validatorInfo = bftVotes.activeValidatorsVoteInfo.find(v =>
+		v.address.equals(newBlockBFTInfo.generatorAddress),
+	);
+	if (!validatorInfo) {
+		throw new Error('Invalid state. activeValidatorsVoteInfo must contain the generator address.');
+	}
+	const heightNotPrevoted = getHeightNotPrevoted(bftVotes);
+
+	const minPrecommitHeight = Math.max(
+		validatorInfo.minActiveHeight,
+		heightNotPrevoted + 1,
+		validatorInfo.largestHeightPrecommit + 1,
+	);
+	let hasPrecommitted = false;
+
+	for (const blockBFTInfo of bftVotes.blockBFTInfos) {
+		if (blockBFTInfo.height < minPrecommitHeight) {
+			break;
+		}
+		const params = await paramsCache.getParameters(blockBFTInfo.height);
+		if (blockBFTInfo.prevoteWeight >= params.prevoteThreshold) {
+			const bftValidator = params.validators.find(v =>
+				v.address.equals(blockBFTInfo.generatorAddress),
+			);
+			if (!bftValidator) {
```

Comment:
Also here, if there is no entry, we should assume BFT weight 0 (maybe a previous active delegate was standby delegate in a previous round).

Reply:
Considering the standby delegate is not part of the validators as above, shouldn't this logic changed to `continue`?
`blockBFTInfos` will include the block from standby delegate?

Reply:
We could change the logic to use `continue` here but it should not make a difference as this case should never occur.

As I mentioned, if a delegate is a standby delegate in a round, then they are now counted as not active and therefore `validatorInfo.minActiveHeigh` is the first height of the round since when they are continuosly active delegate. So we should never check this logic for a height where `blockBFTInfo.generatorAddress` is standby delegate.

Reply:
As discussed, we keep it as is.